### PR TITLE
Fingerprint regeneration based on availability of optional arguments

### PR DESCRIPTION
**Pull request by Anilm3**
This PR introduces support for partially generating fingerprints when some of their arguments are optional and regenerating them as data becomes available. To achieve this, the fields of certain fragments (endpoint and session) are cached as they are generated in order to ensure that only new fields are generated as new data becomes available. In addition, the base processor algorithm is now capable of determining at compile time whether a processor has optional parameters and using that information at runtime to determine if the processor needs to be reevaluated with new data.

With regards to existing fragments, the following changes have been made:

- The body hash field of the endpoint fingerprint is now optional.
- All fields of the session fingerprint are now optional.

Related Jira: [APPSEC-54299]

[APPSEC-54299]: https://datadoghq.atlassian.net/browse/APPSEC-54299?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by estringana — APPROVED** [review]

**Review by cataphract — APPROVED** [review]

File: src/processor/base.hpp [hunk 1]
```diff
@@ -33,15 +33,46 @@ struct processor_mapping {
...
     std::pair<ddwaf_object, object_store::attribute> (Class::*)(Args...) const);
 
+template <typename T, typename... Ts> constexpr std::size_t count_optionals()
+{
+    if constexpr (sizeof...(Ts) == 0) {
+        return static_cast<std::size_t>(is_optional_argument<T>::value);
+    } else {
+        return is_optional_argument<T>::value + count_optionals<Ts...>();
+    }
+}
```
[the remark this answers is not in the document]
Comment (Anilm3) [reply src/processor/base.hpp:64]: I never remember to use these... thanks

File: src/processor/base.hpp [hunk 2]
```diff
@@ -91,34 +122,69 @@ template <typename Self> class structured_processor : public base_processor {
...
-            if (!resolve_arguments(
-                    mapping, store, args, std::make_index_sequence<func_traits::nargs>{})) {
-                continue;
+            tuple_type args;
+            auto arg_count = resolve_arguments(
+                mapping, store, args, std::make_index_sequence<func_traits::nargs>{});
+            if constexpr (is_tuple_with_optional<tuple_type>) {
+                // If there are no new optional arguments, or no arguments at all, skip
+                if (arg_count.all == 0 || (arg_count.all == cache.evaluated[i].all &&
+                                              arg_count.optional == cache.evaluated[i].optional)) {
```
[the remark this answers is not in the document]
Comment (Anilm3) [reply src/processor/base.hpp:162]: In theory a processor receiving ephemeral arguments should return a new ephemeral object and indeed the cache would not be updated as the count currently relies on an increasing count, however I intentionally kept the fingerprints non-ephemeral as it is meaningless for a request to have multiple fingerprints of the same type, so the first one generated would be the request fingerprint. In practice we also don't really use ephemeral arguments with fingerprints so this odd behaviour wouldn't be expected anyway.

File: src/processor/fingerprint.cpp [hunk 3]
```diff
@@ -51,213 +53,42 @@ struct string_buffer {
...
+    auto buffer = std::string{str};
 
-template <typename... Generators>
-ddwaf_object generate_fragment(std::string_view header, Generators... generators)
-{
-    const std::size_t total_length = header.size() + 1 + generate_fragment_length(generators...);
+    // By initialising the cow_string with the underlying data
+    // contained within the std::string, we ensure a new one won't be
+    // allocated once the string is modified.
+    cow_string str_lc{buffer.data(), buffer.size()};
```
[the remark this answers is not in the document]
Comment (Anilm3) [reply src/processor/fingerprint.cpp:86]: Indeed, I might need to have a separate type for non-cow strings, it involves a bit of refactoring perhaps. I'll need to think about it but I will tackle this in a separate PR.

File: src/processor/base.hpp [hunk 4]
```diff
@@ -33,15 +33,46 @@ struct processor_mapping {
...
+        return is_optional_argument<T>::value + count_optionals<Ts...>();
+    }
+}
+
+template <typename T> struct tuple_optionals_trait;
+
+template <typename... Ts> struct tuple_optionals_trait<std::tuple<Ts...>> {
+    static constexpr std::size_t count = count_optionals<Ts...>();
+    static constexpr bool value = count > 0;
+};
```
[the remark this answers is not in the document]
Comment (Anilm3) [reply src/processor/base.hpp:71]: Sure, although I still need to add the count inside the struct.

**Pull request merged**